src/main/webapp/app/complaints/complaints-for-students/complaints-student-view.component.ts

```diff
@@ -0,0 +1,160 @@
+import { Component, Input, OnInit } from '@angular/core';
+import { Exercise } from 'app/entities/exercise.model';
+import { Complaint, ComplaintType } from 'app/entities/complaint.model';
+import { ComplaintService } from 'app/complaints/complaint.service';
+import { StudentParticipation } from 'app/entities/participation/student-participation.model';
+import { Result } from 'app/entities/result.model';
+import { ActivatedRoute } from '@angular/router';
+import { Course } from 'app/entities/course.model';
+import { ArtemisServerDateService } from 'app/shared/server-date.service';
+import { Exam } from 'app/entities/exam.model';
+import { AccountService } from 'app/core/auth/account.service';
+import { Submission } from 'app/entities/submission.model';
+import { filter } from 'rxjs/operators';
+import dayjs from 'dayjs';
+
+@Component({
+    selector: 'jhi-complaint-student-view',
+    templateUrl: './complaints-student-view.component.html',
+})
+export class ComplaintsStudentViewComponent implements OnInit {
+    @Input() exercise: Exercise;
+    @Input() participation: StudentParticipation;
+    @Input() result: Result;
+    @Input() exam: Exam;
+    // flag to indicate exam test run. Default set to false.
+    @Input() testRun = false;
+
+    submission: Submission;
+    complaint: Complaint;
+    course?: Course;
+    // Indicates what type of complaint is currently created by the student. Undefined if the student didn't click on a button yet.
+    formComplaintType?: ComplaintType;
+    // the number of complaints that the student is still allowed to submit in the course. this is used for disabling the complain button.
+    numberOfAllowedComplaints = 0;
+    isCurrentUserSubmissionAuthor: boolean;
+    isExamMode: boolean;
+    showComplaintsSection: boolean;
+    timeOfFeedbackRequestValid: boolean;
+    timeOfComplaintValid: boolean;
+
+    ComplaintType = ComplaintType;
+
+    constructor(
+        private complaintService: ComplaintService,
+        private activatedRoute: ActivatedRoute,
+        private serverDateService: ArtemisServerDateService,
+        private accountService: AccountService,
+    ) {}
+
+    /**
+     * Loads the number of allowed complaints and feedback requests
+     */
+    ngOnInit(): void {
+        this.course = this.exercise.course;
+        this.isExamMode = this.exam != undefined;
+        if (this.participation?.id && this.participation.results && this.participation.results.length > 0) {
+            // Make sure results and participation are connected
+            this.result = this.participation.results[0];
+            this.result.participation = this.participation;
+        }
+        if (this.participation.submissions && this.participation.submissions.length > 0) {
+            this.submission = this.participation.submissions[0];
+        }
+        // for normal exercises we track the number of allowed complaints
```

Comment:
what is a normal exercise?

Reply:
Not my comment but I think this refers to course exercises (vs exam exercises). I'll specify it.